akka-stream/src/main/scala/akka/stream/impl/Throttle.scala

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl
+
+import akka.stream.ThrottleMode.{ Enforcing, Shaping }
+import akka.stream.impl.fusing.GraphStages.SimpleLinearGraphStage
+import akka.stream.stage._
+import akka.stream._
+
+import scala.concurrent.duration.{ FiniteDuration, _ }
+import scala.util.control.NonFatal
+
+/**
+ * INTERNAL API
+ */
+private[stream] class Throttle[T](cost: Int,
+                                  per: FiniteDuration,
+                                  maximumBurst: Int,
+                                  costCalculation: (T) ⇒ Int,
+                                  mode: ThrottleMode)
+  extends SimpleLinearGraphStage[T] {
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) {
+    var willStop = false
+    var lastTokens: Long = 0
+    var previousTime: Long = 0
+
+    val speed = calculateSpeed(cost, per)
+    val timerName: String = "ThrottleTimer"
+
+    var currentElement: Option[T] = None
+    def calculateElementCost(elem: T): Long = scale(costCalculation(elem))
+
+    setHandler(in, new InHandler {
+      val scaledMaximumBurst = scale(maximumBurst)
+
+      override def onUpstreamFinish(): Unit =
+        if (isAvailable(out) && isTimerActive(timerName)) willStop = true
+        else completeStage()
```

Comment:
won't this cause trouble when I'd `(1, per = 1.hour)`? It'll only propagate the completion event downstream after one hour has passed. I think we need to terminate eagerly once upstream finishes.